modules/images/webp-uploads/fallback.js

```diff
@@ -4,21 +4,49 @@ window.wpPerfLab = window.wpPerfLab || {};
 	window.wpPerfLab.webpUploadsFallbackWebpImages = function( media ) {
 		for ( var i = 0; i < media.length; i++ ) {
 			try {
-				if ( ! media[i].media_details.sources || ! media[i].media_details.sources['image/jpeg'] ) {
-					continue;
-				}
-
-				var ext = media[i].media_details.sources['image/jpeg'].file.match( /\.\w+$/i );
-				if ( ! ext || ! ext[0] ) {
-					continue;
-				}
 
-				var images = document.querySelectorAll( 'img.wp-image-' + media[i].id );
+				var images = document.querySelectorAll( 'img.wp-image-' + media[ i ].id ),
+					media_sources = media[ i ].media_details.sources;
 				for ( var j = 0; j < images.length; j++ ) {
-					images[j].src = images[j].src.replace( /\.webp$/i, ext[0] );
-					var srcset = images[j].getAttribute( 'srcset' );
+
+					if ( ! media_sources || ! media_sources['image/jpeg'] ) {
+						continue;
+					}
+
+					var srcset = images[ j ].getAttribute( 'srcset' ),
+						sizes = media[ i ].media_details.sizes,
+						sizes_keys = Object.keys( sizes ),
+						flag = true;
+
+					// If a full image is present in srcset, it should be updated.
+					if ( srcset && media_sources['image/webp'] ) {
+						srcset = srcset.replace( media_sources['image/webp'].file, media_sources['image/jpeg'].file );
+					}
```

Comment:
This code makes sense the way our REST API response is currently structured, however I just checked that again and found out our REST API implementation is inconsistent with core's:
* WordPress core returns the data for the `full` image within `media_details.sizes` as opposed to a separate field. When you review the WordPress core implementation of the REST endpoint, you will see that it adds the `full` image data to `media_details.sizes`. This is different from how the metadata itself is structured, but we should stick to the convention that core has established here.
* So we should probably monitor this in our REST API implementation and only provide the `full` image's `sources` within that value too, instead of `media_details.sources`.
* Let's adjust our implementation in the `rest-api.php` file so that it also includes the correct `sources` for the `full` image size.

Once we have made that change, we will no longer need this extra call here. We can simply loop through `sizes_keys` like below, because that will then include the `full` size as expected. It should notably simplify the logic here.

Reply:
Thank you for the review. This feedback is addressed.

Additional to this I found one case in which the REST API endpoint does not return `sizes`. If you upload an image with 100X100 size then we don't get `sizes` in `media_details` to tackle this case:

- Check if `sources` are present in `media_details` then we don't have `sizes` for that image in that case update image `src` only and break the loop https://github.com/WordPress/performance/blob/fix/426-enhance-js-replacement-mechanism/modules/images/webp-uploads/fallback.js#L19-L24
- We have unset `sources` if we have the full size in `sizes` https://github.com/WordPress/performance/blob/fix/426-enhance-js-replacement-mechanism/modules/images/webp-uploads/rest-api.php#L49

Please review it and share your feedback for this implementation.

Reply:
@mukeshpanchal27 Ah, good point. Sorry, I only read this after my review. I will update my one comment below to reflect this, as I agree with your observation.